appinventor/blocklyeditor/src/generators/yail/control.js

```diff
@@ -93,6 +93,50 @@ Blockly.Yail['controls_forEach'] = function() {
          + listCode + Blockly.Yail.YAIL_CLOSE_COMBINATION;
 };
 
+Blockly.Yail['controls_for_each_dict'] = function() {
+  var yail = Blockly.Yail;
+  var prefix = Blockly.usePrefixInYail ? 'local_' : '';
+  var keyName = yail.YAIL_LOCAL_VAR_TAG + prefix + this.getFieldValue('KEY');
+  var valueName = yail.YAIL_LOCAL_VAR_TAG + prefix + this.getFieldValue('VALUE');
+
+  var dictionaryCode = yail.valueToCode(this, 'DICT', yail.ORDER_NONE);
+  var bodyCode = yail.statementToCode(this, 'DO') || yail.YAIL_FALSE;
+
+  var loopIndexName = 'item';
+  var loopIndexCommandAndName = yail.getVariableCommandAndName(loopIndexName);
+  loopIndexName = loopIndexCommandAndName[1];
+  var getListCode = loopIndexCommandAndName[0] + loopIndexName +
+      yail.YAIL_CLOSE_COMBINATION;
+  var getKeyCode = yail.YAIL_CALL_YAIL_PRIMITIVE + 'yail-list-get-item' + yail.YAIL_SPACER
```

Comment:
For the code generation, you can combine the two variable lets into a single let and get rid of the begin, for example:

```scheme
(foreach $item (let (($key ...) ($value ...)) ...body...) ...dict...)
```

In your example code generation, I don't see the value of the `dict` in the code, which makes me think that your test is failing because it has nothing to iterate over. I'd have to test more. You can also write some unit tests in Java for testing the expressions to make sure they work. See buildserver/tests/.../YailEvalTest.java. In particular, see the test starting around line 546.